pkg/flow/internal/controller/config_argument.go

```diff
@@ -0,0 +1,122 @@
+package controller
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/grafana/agent/pkg/river/ast"
+	"github.com/grafana/agent/pkg/river/vm"
+)
+
+type ArgumentConfigNode struct {
+	label         string
+	nodeID        string
+	componentName string
+
+	mut          sync.RWMutex
+	block        *ast.BlockStmt // Current River blocks to derive config from
+	eval         *vm.Evaluator
+	defaultValue any
+	optional     bool
+}
+
+var _ BlockNode = (*ArgumentConfigNode)(nil)
+
+// NewArgumentConfigNode creates a new ArgumentConfigNode from an initial ast.BlockStmt.
+// The underlying config isn't applied until Evaluate is called.
+func NewArgumentConfigNode(block *ast.BlockStmt, globals ComponentGlobals) *ArgumentConfigNode {
+	return &ArgumentConfigNode{
+		label:         block.Label,
+		nodeID:        BlockComponentID(block).String(),
+		componentName: block.GetBlockName(),
+
+		block: block,
+		eval:  vm.New(block.Body),
+	}
+}
+
+type argumentBlock struct {
+	Optional bool `river:"optional,attr,optional"`
+	Default  any  `river:"default,attr,optional"`
+}
+
+// Evaluate implements BlockNode and updates the arguments for the managed config block
+// by re-evaluating its River block with the provided scope. The managed config block
+// will be built the first time Evaluate is called.
+//
+// Evaluate will return an error if the River block cannot be evaluated or if
+// decoding to arguments fails.
+func (cn *ArgumentConfigNode) Evaluate(scope *vm.Scope) error {
+	cn.mut.Lock()
+	defer cn.mut.Unlock()
+
+	var argument argumentBlock
+	if err := cn.eval.Evaluate(scope, &argument); err != nil {
+		return fmt.Errorf("decoding River: %w", err)
+	}
+
+	cn.defaultValue = argument.Default
+	cn.optional = argument.Optional
+
+	if argument.Optional {
+		return nil
+	}
+
+	args := Arguments(scope)
+	if args != nil {
+		if _, found := (args)[cn.label]; found {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("missing required argument %q to module", cn.label)
```

Comment:
I would like for us to find a way to avoid having ArgumentConfigNode to introspect the scope to validate whether required arguments are provided; it feels complex and error prone.

Reply:
This has been addressed using the new module argument cache after evaluate is called at a higher level.